File: prep/implementation/main_iterations/iteration_cpp/connected_components_dfs.cpp
#include <algorithm>
#include <iostream>
#include <set>
#include <vector>

using namespace std;

struct Vertex {
  set<int> vs;

  void add_edge(int v) { this->vs.insert(v); }
  const vector<int> get_adjacent_vertices() const {
    vector<int> neighbours;

    for (int v : this->vs) {
      neighbours.push_back(v);
    }

    return neighbours;
  }
};

class Graph {
private:
  int sz;
  vector<Vertex> vertices;

public:
  Graph(int sz) : sz(sz) {
    for (int i = 0; i < sz; i++) {
      Vertex v;
      this->vertices.emplace_back(v);
    }
  }

  void add_edge(int from, int to) {
    if (from < 0 || from >= this->sz || to < 0 || to >= this->sz) {
      throw "illegal vertex";
    }

    this->vertices[from].add_edge(to);
  }

  const vector<int> get_adjacent_vertices(int v) const {
    if (v < 0 || v >= this->sz) {
      throw "illegal vertex";
    }

    return this->vertices[v].get_adjacent_vertices();
  }

  int size() const { return this->sz; }
};

void dfs(Graph &g, vector<bool> &visited, int curr_vertex, int cc_id, int v1,
         int &v1_id, int v2, int &v2_id) {
  if (visited[curr_vertex]) {
    return;
  }

  visited[curr_vertex] = true;

  if (curr_vertex == v1) {
    v1_id = cc_id;
  }

  if (curr_vertex == v2) {
    v2_id = cc_id;
  }

  for (int neighbour : g.get_adjacent_vertices(curr_vertex)) {
    if (!visited[neighbour]) {
      dfs(g, visited, neighbour, cc_id, v1, v1_id, v2, v2_id);
    }
  }
}

bool is_connected(Graph &g, int from, int to) {
  vector<bool> visited(g.size());
  fill(visited.begin(), visited.end(), false);

  int cc_id = 0;
  int from_id = -1, to_id = -1;
  for (int i = 0; i < g.size(); i++) {
    if (!visited[i]) {
      dfs(g, visited, i, cc_id, from, from_id, to, to_id);
    }
    cc_id++;
  }

  return from_id == to_id;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int n, m, from, to, nq;
  cin >> n >> m;

  Graph g(n);
  for (int i = 0; i < m; i++) {
    cin >> from >> to;
    g.add_edge(from, to);
  }

  cin >> nq;
  while (nq--) {
    cin >> from >> to;
    cout << (is_connected(g, from, to) ? "yes" : "no") << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/main_reference_implementation/ConnectedComponentsDFS.java
import java.util.*;

public class ConnectedComponentsDFS {
  static interface Graph {
    void addEdge(int from, int to);
    List<Integer> getAdjacentVertices(int v);
    int size();
  }

  static class AdjacencySet implements Graph {
    static class Vertex {
      Set<Integer> vs;

      Vertex() { this.vs = new HashSet<>(); }
      void addEdge(int v) { this.vs.add(v); }

      List<Integer> getAdjacentVertices() {
        List<Integer> ns = new ArrayList<>();
        for (int v : this.vs) {
          ns.add(v);
        }

        return ns;
      }
    }

    private List<Vertex> vertices;
    private int size;

    AdjacencySet(int size) {
      this.size = size;
      this.vertices = new ArrayList<>(size);

      for (int i = 0; i < size; i++) {
        this.vertices.add(new Vertex());
      }
    }

    @Override
    public void addEdge(int from, int to) {
      if (from < 0 || from >= this.size || to < 0 || to >= this.size) {
        throw new IllegalArgumentException("invalid vertex");
      }

      this.vertices.get(from).addEdge(to);
    }

    @Override
    public List<Integer> getAdjacentVertices(int v) {
      if (v < 0 || v >= this.size) {
        throw new IllegalArgumentException("invalid vertex");
      }

      return this.vertices.get(v).getAdjacentVertices();
    }

    @Override
    public int size() {
      return this.size;
    }
  }

  static class IntWrapper {
    int id;
    int ccId;

    IntWrapper(int id, int ccId) {
      this.id = id;
      this.ccId = ccId;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int n = in.nextInt();
      int m = in.nextInt();

      Graph g = new AdjacencySet(n);
      for (int i = 0; i < m; i++) {
        int from = in.nextInt();
        int to = in.nextInt();

        g.addEdge(from, to);
        g.addEdge(to, from);
      }

      int nq = in.nextInt();
      while (nq-- > 0) {
        int v1 = in.nextInt();
        int v2 = in.nextInt();

        System.out.println(dfs(g, v1, v2) ? "yes" : "no");
      }
    }
  }

  private static boolean dfs(Graph g, int v1, int v2) {
    boolean[] visited = new boolean[g.size()];
    int ccId = 0;
    IntWrapper v1Id = new IntWrapper(v1, -1);
    IntWrapper v2Id = new IntWrapper(v2, -1);

    for (int i = 0; i < g.size(); i++) {
      dfs(g, visited, i, ccId, v1Id, v2Id);
      ccId++;
    }

    if (v1Id.ccId == v2Id.ccId) {
      return true;
    }
    return false;
  }

  private static void dfs(Graph g, boolean[] visited, int currVertex, int ccId,
                          IntWrapper v1Id, IntWrapper v2Id) {
    if (visited[currVertex]) {
      return;
    }

    if (currVertex == v1Id.id) {
      v1Id.ccId = ccId;
    }

    if (currVertex == v2Id.id) {
      v2Id.ccId = ccId;
    }

    visited[currVertex] = true;
    for (int neighbour : g.getAdjacentVertices(currVertex)) {
      if (!visited[neighbour]) {
        dfs(g, visited, neighbour, ccId, v1Id, v2Id);
      }
    }
  }
}